### PR TITLE
Volume of the sphere

### LINKED ISSUE #25: Volume of the Sphere Challenge
"""

## Description
Write a Python function that calculates the volume of a sphere.

## **Behavior**
The function should:

1. Take the radius of the sphere (a float) as input.
2. Calculate the volume using the formula: V = (4/3) *π* r^3, where r is the radius.
3. Return the calculated volume (a float).
4. Ensure that the radius should be a positive number. 

## **Examples**

**Example 1**
Input: 5.0
Output: 523.5987755982989

**Example 2**
Input: 2.5
Output: 65.44984694978735

"""

### PR DESCRIPTION
**Pull request by Kuhai9801**
---
name: Volume of The Sphere Challenge
about: A template PR for code review with a checklist
---

## Behavior
Calculates the Volume of a Sphere from a given input radii.

### Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [x] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [x] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [x] Each assumption about an argument is checked with an assertion
  - [x] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
  - _It's ok to have extra helper functions if necessary, like with mergesort_
- [x] The function has correct type annotations
- [x] The function is not called at the top level of the function file 
  - _Recursive solutions **can** call the function from **inside** the function body_

### Do's

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [x] The function's strategy _is not_ described in any docstrings or tests
- [x] Comments explain the _strategy_, **not** the _implementation_
- [x] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] There are no commented lines of code
- [x] There are no `print` statements anywhere
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

### CODE REVIEW

File: solutions/volume_of_the_sphere.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+        >>> volume_of_the_sphere(1.0)
+        4.1887902047863905
+    """
+    # Type validation
+    assert isinstance(radius, (int, float)), "Radius must be a number"
+    assert type(radius) in (int, float), "Radius must be int or float"
+
+    # Value validation
+    assert radius != 0, "Radius cannot be zero"
+    assert radius > 0, "Radius must be positive"
```
Comment (Ammmoe) [inline solutions/volume_of_the_sphere.py]: If you want to raise `ValueError`, you can write:
```py
if radius <= 0:
    raise ValueError("Radius must have a positive value")
```

File: solutions/volume_of_the_sphere.py [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+        >>> volume_of_the_sphere(5.0)
+        523.5987755982989
+        >>> volume_of_the_sphere(2.5)
+        65.44984694978736
+        >>> volume_of_the_sphere(1.0)
+        4.1887902047863905
+    """
+    # Type validation
+    assert isinstance(radius, (int, float)), "Radius must be a number"
+    assert type(radius) in (int, float), "Radius must be int or float"
```
Comment (Ammmoe) [inline solutions/volume_of_the_sphere.py]: You can `delete` this line. Line 40 already raises `AssertionError` if the radius is not an int or a float.

File: solutions/tests/test_volume_of_the_sphere.py [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+        """Test volume calculation with integer input."""
+        self.assertAlmostEqual(volume_of_the_sphere(3), (4 / 3) * pi * 27)
+
+    def test_raises_type_error_with_string_input(self):
+        """Test TypeError raised when input is a string."""
+        with self.assertRaises(AssertionError):
+            volume_of_the_sphere("5")
+
+    def test_raises_type_error_with_list_input(self):
+        """Test TypeError raised when input is a list."""
```
Comment (Ammmoe) [inline solutions/tests/test_volume_of_the_sphere.py:53]: Please change `TypeError` to `AssertionError` as you raised an AssertionError in your code.

Or please raise `TypeError` in your code.

File: solutions/tests/test_volume_of_the_sphere.py [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+    def test_volume_with_radius_one(self):
+        """Test volume calculation with radius 1.0."""
+        self.assertAlmostEqual(volume_of_the_sphere(1.0), 4.1887902047863905)
+
+    def test_volume_with_integer_radius(self):
+        """Test volume calculation with integer input."""
+        self.assertAlmostEqual(volume_of_the_sphere(3), (4 / 3) * pi * 27)
+
+    def test_raises_type_error_with_string_input(self):
+        """Test TypeError raised when input is a string."""
```
Comment (Ammmoe) [inline solutions/tests/test_volume_of_the_sphere.py:48]: Please change `TypeError` to `AssertionError` as you raised an AssertionError in your code.

Or please raise `TypeError` in your code.

File: solutions/tests/test_volume_of_the_sphere.py [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+        with self.assertRaises(AssertionError):
+            volume_of_the_sphere("5")
+
+    def test_raises_type_error_with_list_input(self):
+        """Test TypeError raised when input is a list."""
+        with self.assertRaises(AssertionError):
+            volume_of_the_sphere([5])
+
+    def test_raises_value_error_with_zero_radius(self):
+        """Test ValueError raised when radius is zero."""
```
Comment (Ammmoe) [inline solutions/tests/test_volume_of_the_sphere.py:58]: Please raise `ValueError` in your function.

File: solutions/volume_of_the_sphere.py [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+
+    # Value validation
+    assert radius != 0, "Radius cannot be zero"
+    assert radius > 0, "Radius must be positive"
+
+    return (4 / 3) * math.pi * (radius**3)
```
Comment (Ammmoe) [inline solutions/volume_of_the_sphere.py]: This is just a `suggestion` because you have to test with a lot of decimal places.

If you want to round the decimal places to let's say 2 positions, you can write it like this:
```py
return round((4 / 3) * math.pi * (radius**3), 2)
```

Or if you want to round to 3 decimal places, you can:
```py
return round((4 / 3) * math.pi * (radius**3), 3)
```

File: solutions/volume_of_the_sphere.py [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+    - volume_of_the_sphere: returns the volume of a sphere given its radius
+
+Created on 2025-01-07
+@author: Cyne Jarvis J. Zarceno
+"""
+
+import math
+
+
+def volume_of_the_sphere(radius: float) -> float:
```
Comment (Ammmoe) [inline solutions/volume_of_the_sphere.py:17]: You're correct in this code as float type also accepts integers. 

However, I heard that python >= 3.10 allows the following so you can try:
```py
def volume_of_the_sphere(radius: int | float) -> int | float:
```

File: solutions/tests/test_volume_of_the_sphere.py (hunk 3 above)
Comment (1sgtpepper) [reply solutions/tests/test_volume_of_the_sphere.py:53]: removed it to adhere to project description, I forgot to include it in the function code too. Thanks

File: solutions/tests/test_volume_of_the_sphere.py (hunk 4 above)
Comment (1sgtpepper) [reply solutions/tests/test_volume_of_the_sphere.py:48]: removed it to adhere to project description, I forgot to include it in the function code too. Thanks

File: solutions/volume_of_the_sphere.py (hunk 2 above)
Comment (1sgtpepper) [reply solutions/volume_of_the_sphere.py]: Thank you for this! Changed it.

File: solutions/volume_of_the_sphere.py (hunk 1 above)
Comment (1sgtpepper) [reply solutions/volume_of_the_sphere.py]: Thank you!

File: solutions/volume_of_the_sphere.py (hunk 6 above)
Comment (1sgtpepper) [reply solutions/volume_of_the_sphere.py]: Thank you for this suggestion! Since people usually use 2 decimal places, I'll stick with the first suggestion.

File: solutions/volume_of_the_sphere.py [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+    """Returns the volume of a sphere given its radius
```
Comment (Ammmoe) [inline solutions/volume_of_the_sphere.py]: This is a function description so it'd be better to use `Calculates` instead of `Returns`.

File: solutions/tests/test_volume_of_the_sphere.py (hunk 5 above)
Comment (1sgtpepper) [reply solutions/tests/test_volume_of_the_sphere.py:58]: Thank you for pointing this out! This is appreciated!

File: solutions/volume_of_the_sphere.py (hunk 8 above)
Comment (1sgtpepper) [reply solutions/volume_of_the_sphere.py]: Thanks, will get to it right away

File: solutions/volume_of_the_sphere.py [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+
+    Parameter:
+        radius (float): Positive real numbers.
```
Comment (Ammmoe) [inline solutions/volume_of_the_sphere.py:22]: You can also include the purpose and the assumptions of the argument:
```py
Parameter: radius (int | float)
    - the radius of the sphere to calculate its volume
    - must be a positive real number
```

File: solutions/volume_of_the_sphere.py [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+
+    Returns -> float: The volume of the sphere calculated using the formula V = (4/3)πr³
+
+    Raises:
+        AssertionError: If radius is not a number
+        AssertionError: If radius is not int or float
+        ValueError: If radius is zero
+        ValueError: If radius is negative
```
Comment (Ammmoe) [inline solutions/volume_of_the_sphere.py]: You can reformatted it like this to be more efficient:
```py
Raises:
    AssertionError: if radius is not an integer or a float
    ValueError: if radius is zero or negative
```

Comment (1sgtpepper) [reply solutions/volume_of_the_sphere.py]: Did not think of that! thanks.

File: solutions/volume_of_the_sphere.py [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+import math
+
+
+def volume_of_the_sphere(radius: float) -> float:
+    """Returns the volume of a sphere given its radius
+
+    Parameter:
+        radius (float): Positive real numbers.
+
+    Returns -> float: The volume of the sphere calculated using the formula V = (4/3)πr³
```
Comment (Ammmoe) [inline solutions/volume_of_the_sphere.py]: Maybe it's just me but the formula part sounds like a `strategy` so I think you do not include it here and add it to your implementation comments.
```py
Returns -> int | float
    - returns the calculated volume of the sphere
```

File: solutions/volume_of_the_sphere.py (hunk 9 above)
Comment (1sgtpepper) [reply solutions/volume_of_the_sphere.py:22]: Will take this into account. However I think float should be enough since it also encompasses integers(can you clarify this for me or is this for clarity?).  Thanks

File: solutions/volume_of_the_sphere.py (hunk 11 above)
Comment (1sgtpepper) [reply solutions/volume_of_the_sphere.py]: Changed it!

File: solutions/volume_of_the_sphere.py [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+        4.1887902047863905
+    """
+    # Type validation
+    assert isinstance(radius, (int, float)), "Radius must be a number"
+    assert type(radius) in (int, float), "Radius must be int or float"
+
+    # Value validation
+    assert radius != 0, "Radius cannot be zero"
+    assert radius > 0, "Radius must be positive"
+
```
Comment (Ammmoe) [inline solutions/volume_of_the_sphere.py:46]: You can insert your strategy here with a comment: 
```py
# Calculate the volume of the sphere using the formula V = (4/3)πr³
```

**Review by Ammmoe — CHANGES_REQUESTED** [review]
Hi @Kuhai9801 I added some comments to your pull request files. It includes suggestions for the function behaviour, docstring, strategy, and implementation.

Please make some changes for the suggestions you prefer and please discuss by replying the comments for the suggestions you do not prefer.

File: solutions/volume_of_the_sphere.py (hunk 12 above)
Comment (1sgtpepper) [reply solutions/volume_of_the_sphere.py:46]: Done!

File: solutions/volume_of_the_sphere.py (hunk 7 above)
Comment (1sgtpepper) [reply solutions/volume_of_the_sphere.py:17]: Thank you for this(interesting implementation btw), but I would like to stick with the previous code for the meantime, will try to study this too(python versions etc.)

File: solutions/tests/test_volume_of_the_sphere.py (hunk 3 above)
Comment (Ammmoe) [reply solutions/tests/test_volume_of_the_sphere.py:53]: Hi @Kuhai9801 Sorry. You do not need to remove the `Defensive tests`. This is a good practice.
You can include the previous function with the `AssertionError` in the unit test instead of mentioning the `TypeError` which is not raised.
```py
def test_raises_assertion_error_with_string_input(self):
    """Test AssertionError raised when input is a string."""
    with self.assertRaises(AssertionError):
        volume_of_the_sphere("5")

def test_raises_assertion_error_with_list_input(self):
    """Test AssertionError raised when input is a list."""
    with self.assertRaises(AssertionError):
        volume_of_the_sphere([5])
```

File: solutions/tests/test_volume_of_the_sphere.py (hunk 4 above)
Comment (Ammmoe) [reply solutions/tests/test_volume_of_the_sphere.py:48]: Hi @Kuhai9801 Sorry. You do not need to remove the `Defensive tests`. This is a good practice.
You can include the previous function with the `AssertionError` in the unit test instead of mentioning the `TypeError` which is not raised.
```py
def test_raises_assertion_error_with_string_input(self):
    """Test AssertionError raised when input is a string."""
    with self.assertRaises(AssertionError):
        volume_of_the_sphere("5")

def test_raises_assertion_error_with_list_input(self):
    """Test AssertionError raised when input is a list."""
    with self.assertRaises(AssertionError):
        volume_of_the_sphere([5])
```

File: solutions/volume_of_the_sphere.py (hunk 9 above)
Comment (Ammmoe) [reply solutions/volume_of_the_sphere.py:22]: Yes I think floats also includes integers as well.

However, the main reason your code accepts integers is because you accepts integers in your defensive assertion:
```py
assert isinstance(radius, (int, float)), "Radius must be a number"
```

File: solutions/volume_of_the_sphere.py [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+        AssertionError: if radius is not an integer or a float
+        ValueError: if radius is zero or negative
+
+    Examples:
+        >>> volume_of_the_sphere(5.0)
+        523.5987755982989
+        >>> volume_of_the_sphere(2.5)
+        65.44984694978736
+        >>> volume_of_the_sphere(1.0)
+        4.1887902047863905
```
Comment (Ammmoe) [inline solutions/volume_of_the_sphere.py]: Hi @Kuhai9801 ! Since you rounded the calculated volume. You need to change the answers here as well in order to pass the doctests.

**Review by Ammmoe — CHANGES_REQUESTED** [review]
@Kuhai9801 Things look great overall. However please change the followings before I merge your pull request:

1. To add back the defensive assertions for invalid input (strings and lists) to the unittest
2. To resolve the failing doctests in your function docstring

**Comment by 1sgtpepper** [comment]
It seems that we had a misunderstanding with my comment, in my first plan, I was supposed to only check for value errors(which I also forgot, this is why I commented stuff about project description). In addition, I also made some modifications for the Type Validation for your suggestions in the test, and made changes according to your feedbacks. Thank you for your suggestions.

File: solutions/tests/test_volume_of_the_sphere.py [hunk 14]
```diff
@@ -13,14 +13,19 @@
...
-import os
 
-sys.path.append(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
 from solutions.volume_of_the_sphere import volume_of_the_sphere
 
+# Add parent directory to Python path
+sys.path.append(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
+current_dir = os.path.dirname(os.path.abspath(__file__))
+parent_dir = os.path.dirname(current_dir)
+sys.path.insert(0, parent_dir)
```
Comment (Ammmoe) [inline solutions/tests/test_volume_of_the_sphere.py]: Hi @Kuhai9801 Could you please remove lines 25, 26, and 27? I think these three lines are doing the same thing as line 24.

**Review by Ammmoe — APPROVED** [review]
Thank you @Kuhai9801 for your patience in modifying the changes. It looks great and ready to be merged.
